src/workerd/io/tracer.c++

```diff
@@ -203,6 +203,34 @@ void WorkerTracer::addSpan(CompleteSpan&& span) {
     return;
   }
 
+  // To report I/O time, we need the IOContext to still be alive.
+  // weakIoContext is only none if we are tracing via RPC (in this case span times have already been
+  // adjusted) or if we failed to transmit an Onset event (in that case we'll get an error based on
+  // missing topLevelInvocationSpanContext right after).
+  if (weakIoContext != kj::none) {
+    auto& weakIoCtx = KJ_ASSERT_NONNULL(weakIoContext);
+    weakIoCtx->runIfAlive([&span](IoContext& context) { span.endTime = context.now(); });
+    if (!weakIoCtx->isValid()) {
+      // This can happen if we start a customEvent from this event and cancel it after this IoContext
+      // gets destroyed. In that case we no longer have an IoContext available and can't get the
+      // current time, but the outcome timestamp will have already been set. Since the outcome
+      // timestamp is "late enough", simply use that.
+      if (completeTime != kj::UNIX_EPOCH) {
+        span.endTime = completeTime;
+      } else {
+        // Otherwise, we can't actually get an end timestamp that makes sense. Report a zero-duration
+        // span and log a warning (or fail assert in test mode).
+        span.endTime = span.startTime;
+        if (isPredictableModeForTest()) {
+          KJ_FAIL_ASSERT("reported span after IoContext was deallocated", span.operationName);
+        } else {
+          LOG_WARNING_PERIODICALLY(
+              "reported span after IoContext was deallocated", span.operationName);
+        }
+      }
```

Comment:
```suggestion
      if (completeTime != kj::UNIX_EPOCH) {
        span.endTime = completeTime;
      } else {
        // Otherwise, we can't actually get an end timestamp that makes sense. Report a zero-duration
        // span and log a warning (or fail assert in test mode).
        span.endTime = span.startTime;
      }
      if (isPredictableModeForTest()) {
        KJ_FAIL_ASSERT("reported span after IoContext was deallocated", span.operationName);
      } else {
        LOG_WARNING_PERIODICALLY(
            "reported span after IoContext was deallocated", span.operationName);
      }
```
Closing a span without a valid IoContext is a bug either way, even if the span was closed after the event has completed successfully.

Reply:
Well, this currently happens for spans like jsRpcSession, and likely other spans that are attached to a promise that we call `IoContext::addTask` on: https://github.com/cloudflare/workerd/blob/main/src/workerd/api/http.c%2B%2B#L2470
The tests would fail based on jsRpcSession with the proposed change. This happens when the IoContext gets canceled, so when the request is not successful (as described in the comment).
We could try to figure out another place to attach the span in this case (it is implicitly attached to the worker variable, but attaching to event doesn't help), or try to somehow end the span when the IoContext gets shut down. This seems difficult though, so perhaps not in scope for this PR?

Reply:
👍